File: java/core/libjoynr/src/main/java/io/joynr/arbitration/KeywordArbitrator.java
package io.joynr.arbitration;

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2013 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import io.joynr.capabilities.CapabilityEntry;
import io.joynr.capabilities.LocalCapabilitiesDirectory;

import java.util.Collection;

import javax.annotation.CheckForNull;

import joynr.types.CustomParameter;
import joynr.types.ProviderQos;
import joynr.types.ProviderQosRequirements;

public class KeywordArbitrator extends Arbitrator {

    private String domain;
    private String interfaceName;
    private String requestedKeyword;

    public KeywordArbitrator(final String domain,
                             String interfaceName,
                             final DiscoveryQos discoveryQos,
                             LocalCapabilitiesDirectory capabilitiesSource,
                             long minimumArbitrationRetryDelay) {
        super(discoveryQos, capabilitiesSource, minimumArbitrationRetryDelay);
        this.domain = domain;
        this.interfaceName = interfaceName;
    }

    @Override
    public void startArbitration() {
        arbitrationStatus = ArbitrationStatus.ArbitrationRunning;
        notifyArbitrationStatusChanged();
        requestedKeyword = discoveryQos.getCustomParameter(ArbitrationConstants.KEYWORD_PARAMETER).toString();
        ProviderQosRequirements providerQosRequirements = new ProviderQosRequirements();

        localCapabilitiesDirectory.getCapabilities(domain,
                                                   interfaceName,
                                                   providerQosRequirements,
                                                   discoveryQos,
                                                   callback);
    }

    @Override
    protected void selectProvider(@CheckForNull final Collection<CapabilityEntry> capabilities) {
        if (capabilities == null || capabilities.size() < 1) {
            arbitrationStatus = ArbitrationStatus.ArbitrationCanceledForever;
            notifyArbitrationStatusChanged();
        } else {
            CapabilityEntry result;
            for (CapabilityEntry capEntry : capabilities) {
                ProviderQos providerQos = capEntry.getProviderQos();

                // If onChange subscriptions are required ignore providers that do not support them
                if (discoveryQos.getProviderMustSupportOnChange() && !providerQos.getSupportsOnChangeSubscriptions()) {
                    continue;
                }

                // Search for a matching keyword parameter
                CustomParameter keywordParameter = findQosParameter(capEntry, ArbitrationConstants.KEYWORD_PARAMETER);
                if (keywordParameter != null) {
                    String currentKeyword = keywordParameter.getValue();
                    if (currentKeyword.equals(requestedKeyword)) {
                        result = capEntry;
                        arbitrationResult.setEndpointAddress(result.getEndpointAddresses());
                        arbitrationResult.setParticipantId(result.getParticipantId());
                        arbitrationStatus = ArbitrationStatus.ArbitrationSuccesful;
                        updateArbitrationResultAtListener();
                        return;
                    }
                }
            }
            arbitrationStatus = ArbitrationStatus.ArbitrationCanceledForever;
            notifyArbitrationStatusChanged();
        }
    }
}

File: cpp/libjoynr/proxy/KeywordArbitrator.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2016 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "joynr/KeywordArbitrator.h"
#include "joynr/DiscoveryQos.h"
#include "joynr/system/IDiscovery.h"
#include "joynr/types/DiscoveryEntry.h"
#include "joynr/exceptions/JoynrException.h"
#include "joynr/TypeUtil.h"

namespace joynr
{

INIT_LOGGER(KeywordArbitrator);

KeywordArbitrator::KeywordArbitrator(const std::string& domain,
                                     const std::string& interfaceName,
                                     const joynr::types::Version& interfaceVersion,
                                     joynr::system::IDiscoverySync& discoveryProxy,
                                     const DiscoveryQos& discoveryQos)
        : ProviderArbitrator(domain, interfaceName, interfaceVersion, discoveryProxy, discoveryQos),
          keyword(discoveryQos.getCustomParameter(DiscoveryQos::KEYWORD_PARAMETER()).getValue())
{
}

std::string KeywordArbitrator::filterDiscoveryEntries(
        const std::vector<joynr::types::DiscoveryEntry>& discoveryEntries)
{
    std::string res;
    for (joynr::types::DiscoveryEntry discoveryEntry : discoveryEntries) {
        types::ProviderQos providerQos = discoveryEntry.getQos();
        JOYNR_LOG_TRACE(logger, "Looping over capabilitiesEntry: {}", discoveryEntry.toString());

        // Search the QosParameters for the keyword field
        std::vector<types::CustomParameter> qosParameters = providerQos.getCustomParameters();
        for (types::CustomParameter parameter : qosParameters) {
            std::string name = parameter.getName();
            if (!(name == DiscoveryQos::KEYWORD_PARAMETER() && keyword == parameter.getValue())) {
                continue;
            } else {
                std::string res = discoveryEntry.getParticipantId();
                JOYNR_LOG_TRACE(logger, "setting res to {}", res);
                notifyArbitrationListener(res);
            }
        }
    }

    if (res.empty()) {
        std::string errorMsg;
        errorMsg = "There was more than one entries in capabilitiesEntries, but none of the "
                   "compatible entries had the correct keyword.";
        JOYNR_LOG_WARN(logger, errorMsg);
        arbitrationError.setMessage(errorMsg);
    }
    return res;
}

} // namespace joynr
